Keccak refactoring with new Cell Manager

leolara opened the pull request:
### Description

[_PR description_]

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

[_explanation_]

<hr>

## How to fill a PR description 

Please give a concise description of your PR.

The target readers could be future developers, reviewers, and auditors. By reading your description, they should easily understand the changes proposed in this pull request.

MUST: Reference the issue to resolve

### Single responsability

Is RECOMMENDED to create single responsibility commits, but not mandatory.

Anyway, you MUST enumerate the changes in a unitary way, e.g.

```
This PR contains:
- Cleanup of xxxx, yyyy
- Changed xxxx to yyyy in order to bla bla
- Added xxxx function to ...
- Refactored ....
```

### Design choices

RECOMMENDED to:
- What types of design choices did you face?
- What decisions you have made?
- Any valuable information that could help reviewers to think critically

ChihChengLiang commented on zkevm-circuits/src/util/cell_manager.rs:
```diff
@@ -162,6 +198,27 @@ pub(crate) trait CellManagerStrategy {
         cell_type: CellType,
     ) -> Cell<F>;
 
+    fn query_cell_with_affinity<F: Field>(
+        &mut self,
+        columns: &mut CellManagerColumns,
+        meta: &mut ConstraintSystem<F>,
+        cell_type: CellType,
+        affinity: Self::Affinity,
+    ) -> Cell<F>;
+
+    fn query_cell_value(
+        &mut self,
+        columns: &mut CellManagerColumns,
+        cell_type: CellType,
+    ) -> CellValueOnly;
+
+    fn query_cell_value_with_affinity(
```
Can we doc these three methods?

leolara replied:
Done

ChihChengLiang commented on zkevm-circuits/src/util/cell_manager_strategy.rs:
```diff
@@ -182,26 +183,60 @@ impl CellManagerStrategy for CMFixedWidthStrategy {
 
-// TODO: This strategy is unfinished.
+#[derive(Debug, Clone)]
 pub(crate) struct CMFixedHeigthStrategy {
```
I noticed a typo here. But we can also leave it as a good first issue.
```suggestion
pub(crate) struct CMFixedHeightStrategy {
```

leolara replied:
Done

ChihChengLiang reviewed (commented):
My first pass, and some quick comments.

leolara commented on zkevm-circuits/src/keccak_circuit.rs:
```diff
@@ -112,7 +113,12 @@ impl<F: Field> SubCircuitConfig<F> for KeccakCircuitConfig<F> {
         let pack_table = array_init::array_init(|_| meta.lookup_table_column());
 
-        let mut cell_manager = CellManager::new(get_num_rows_per_round());
+        // let mut cell_manager = CellManager::new(get_num_rows_per_round());
```
Done

leolara commented on zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:
```diff
@@ -39,6 +43,30 @@ pub(crate) fn get_num_bits_per_base_chi_lookup() -> usize {
 }
 
+pub(crate) trait AssignKeccakRegion {
+    fn assign_keccak_region<F: Field>(&self, region: &mut KeccakRegion<F>, offset: i32, value: F);
```
Done

leolara commented on zkevm-circuits/src/util/cell_manager.rs:
```diff
@@ -74,7 +74,7 @@ impl<F: Field> Cell<F> {
             _column_expression: meta.query_advice(column, Rotation::cur()),
             column,
-            _column_idx,
+            column_idx: _column_idx,
```
Done, also removed _column_expression that is not used

leolara commented on zkevm-circuits/src/util/cell_manager.rs:
```diff
@@ -280,4 +360,8 @@ impl<Stats, S: CellManagerStrategy<Stats = Stats>> CellManager<S> {
         self.strategy.get_stats(&self.columns)
     }
+
+    pub fn get_strategy(&mut self) -> &mut S {
+        &mut self.strategy
+    }
```
I wouldn't do (1) but because in my opinion this feature looks like a keccak specific thing.

Regarding (2), isn't bending the meaning of the trait that cell_manager->deref_mut returns a strategy instead of a cell manager?

ChihChengLiang replied:
For (2), I think deref allows you to call both methods of the cell_manager and methods of the strategy.

But I agree with you on point (1) and that (2) might be confusing. The status quo looks good to me now.

hero78119 replied:
Feel that, if there are many code duplication, then it might be some place to improve.

Notice that, the CellManager is like a just columns warehouse
```
pub(crate) struct CellManager<S: CellManagerStrategy> {
    columns: CellManagerColumns,
    strategy: S,
}
```

and most of columns manipulation logic are now based on strategy S. That's why there are a lot of `bypass logic` on cell manager because we try to call a method on column warehouse. 

How about we swap the role so the strategy is new "cellmanager", then during initialization we just pass a column warehouse to it? Feel it can save lot of code re-implementation

leolara replied:
So, precisely the point of this is to avoid each of the modules to use a different cell manager directly because they will start to diverge, even if we say that they have to use the same trait.

This is the strategy pattern https://en.wikipedia.org/wiki/Strategy_pattern

All users will only communicate with the same struct Cell Manager, the parts that are different behaviour (how cells are placed) will be handle by individual strategies that will implement different behaviours.

This was presented as the design of this feature: https://hackmd.io/_RIuqSZSTp-p8fZ4ohbyVg which was available for feedback for months.

However, for a bad design (from the software engineering POV) of the keccak circuit had made necessary to add some weird stuff like the value queries. Keccak queries the cells all over in functions and such.

The right design for a circuit using the Cell Manager would be to always declare the signals before using them, and from that declaration used them both in the setup and the witness generation. This allow simpler Cell Manager code and more clever optimizations. Like it is done in Chiquito: https://github.com/privacy-scaling-explorations/chiquito/blob/main/src/compiler/cell_manager.rs

I hope this PR help us to get in that direction.

hero78119 replied:
Thanks for the explanation!

A essential different I spotted is, in strategy_patten, `class`, in our case, cellmamager, do not need to implement strategy interface, and class is more like a high level encapsulation of its strategy underlyne. It's functionality shouldn't be one to one mapping. Take [Wi](https://en.wikipedia.org/wiki/Strategy_pattern) example, a `car` is a class, where one of its functionality `brake` will be implemented as different strategy. However, car do not implement brake interface.

In cell manager case, its more like we have a unnessary abstraction to have a `Car` encaptulate another `Car`. From this perspective, does it make sense that every modules just rely on CellManagerStrategy trait, where 2 implementation column/row oriented for it ? (CellManagerStrategy can just rename to CellManager)

leolara replied:
1.

The design was available months ago. It was encouraged in the scrum meetings to go and comment, and it was approved by Edu and CC. Adría and Carlos also checked it.

https://hackmd.io/_RIuqSZSTp-p8fZ4ohbyVg  

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203#issuecomment-1516197888

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203#issuecomment-1516384404

2.

Following your comments I have changed some naming in the code to make more clear the intention. I also have moved out of the strategy the only repeated code, the creation of the Cell.

Additionally, I have added comments to clarify that query_value is deprecated. The configuration and synthesize should share the cells in the same way they share the columns.

3.

If we do what you propose, that would return to the previous status quo, where each circuit has a different cell manager.

4.

Because of the example in the wikipedia you might have inferred that the “Context class” (Car) should implement more features than the strategy interface (IBrake). That is never stated as the intention of the Strategy Pattern. It is just a choice of example.

5.

In this more extended article https://refactoring.guru/design-patterns/strategy

We can see that in its example, the context class (Navigator) only goal from the point of the client is to build a route, and then it wraps different ways of doing it (walking, by car…) in different strategies that implement the same interface.

It is not necessary	or condition or related at all that the context class have other actions beyond the one done in the strategies. It is an interface between the client and the strategies.

6.

The goal of the strategy pattern is NOT to wrap different behaviours in a single class with internal implementer for each behaviour, like Break, Accelerate, etc,,

The goal of the strategy pattern is to give the clients of a problem solver a single point of entry to different strategies to solve that problem.

So in our cases, the point is that all circuits will communicate with only one struct CellManager, and then the strategy used will be transparent to the circuit.

ChihChengLiang replied:
@hero78119 I think @leolara has a point on the strategy design.
Can we maintain the current scope of the PR and move the potential design change to a new issue and PR?

leolara commented on zkevm-circuits/src/keccak_circuit.rs:
```diff
@@ -112,7 +113,12 @@ impl<F: Field> SubCircuitConfig<F> for KeccakCircuitConfig<F> {
+        let mut cell_manager = CellManager::new(CMFixedHeigthStrategy::new(
+            get_num_rows_per_round(),
+            DEFAULT_CELL_TYPE,
+        ));
```
I think this is kind of against the idea that you can create new strategies without modifying the cell manager.

ChihChengLiang replied:
Oh, I wasn't aware of that design choice. Then, keeping the status quo sounds good.

leolara commented on zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:
```diff
@@ -596,7 +647,11 @@ pub(crate) fn keccak<F: Field>(
+                    is_padding.assign_keccak_region(
+                        &mut region,
+                        0,
+                        if padding { F::ONE } else { F::ZERO },
```
I tried doing this, I would have to add a lot of F: From<bool> as trait Field does not include this trait. Do you think this is better? I am not inclined either way to this one. F::from(padding.into()) looks better, but adding From<bool> in a lot of places that does not have anything related to bool could also cause confusion.

ChihChengLiang replied:
> ... adding From in a lot of places that does not have anything related to bool 

Makes sense to me. We can keep the status quo.

ChihChengLiang reviewed (approved):
LGTM.

hero78119 commented on zkevm-circuits/src/util/cell_manager.rs:
```diff
@@ -122,6 +129,27 @@ impl<F: Field> Expr<F> for &Cell<F> {
 }
 
+#[derive(Clone)]
+pub(crate) struct CellValueOnly {
```
It seems that `CellValueOnly` 
- is pretty specific to keccak logic however defined in shared place
- it's just a subset of Cell<F> with quite of field being remove.

Can we just follow the design on old `Cell<F>` https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/keccak_circuit/cell_manager.rs#L35-L43 and extend existing `Cell<F>` with a new `new_value_only` api so we do not implemenet dedicated api for `CellValueOnly`

leolara replied:
The thing with this is that Column would need to be an Option, do you think that is the best way?

I am seeing that actually all fields in Cell could be private, so we could simplify the code, while putting column as Option. I am going to try

hero78119 replied:
I think `Option<Column>` make sense 👍  since we have this use case in keccak

hero78119 reviewed (commented):
First round pass and overall LGTM! most are line by line translation
Leave 2 comments related to simplify the design, please help to see whether make sense  :)

hero78119 commented on zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:
```diff
@@ -39,6 +43,22 @@ pub(crate) fn get_num_bits_per_base_chi_lookup() -> usize {
+pub(crate) trait AssignKeccakRegion {
+    fn assign_keccak_region<F: Field>(&self, region: &mut KeccakRegion<F>, value: F);
+}
+
```
This 2 implementation are exactly the same, maybe can move to `trait` default implementation.

https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations

leolara replied:
So, the implementation uses self.column_idx and self.rotation. So to do a default implementation I would have to create another trait with get_column_idx() and get_rotation that is implemented by CellValueOnly and Cell.

Do you think I should do that?

Also, if what you says about unifying CellValueOnly and Cell works, perhaps we don't need to do anything of this.

hero78119 replied:
You are right, need to introduce getter to have this feature :)

Instead, can seek another option to unify CellValueOnly and Cell 👍

leolara commented:
@hero78119 Please check what I have done, I have eliminated CellValueOnly and encapsulated Cell more.

hero78119 commented:
> @hero78119 Please check what I have done, I have eliminated CellValueOnly and encapsulated Cell more.

looks nice!

> Feel that, if there are many code duplication, then it might be some place to improve.
> 
> Notice that, the CellManager is like a just columns warehouse
> 
> ```
> pub(crate) struct CellManager<S: CellManagerStrategy> {
>     columns: CellManagerColumns,
>     strategy: S,
> }
> ```
> 
> and most of columns manipulation logic are now based on strategy S. That's why there are a lot of `bypass logic` on cell manager because we try to call a method on column warehouse.
> 
> How about we swap the role so the strategy is new "cellmanager", then during initialization we just pass a column warehouse to it? Feel it can save lot of code re-implementation

@leolara would it make sense if we swap the cellmanager/strategy so we can avoid many code duplication, due to calling api on cellmanager which is finally delegate to strategy

leolara commented:
1.

The design was available months ago. It was encouraged in the scrum meetings to go and comment, and it was approved by Edu and CC. Adría and Carlos also checked it.

https://hackmd.io/_RIuqSZSTp-p8fZ4ohbyVg  

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203#issuecomment-1516197888

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203#issuecomment-1516384404

2.

Following your comments I have changed some naming in the code to make more clear the intention. I also have moved out of the strategy the only repeated code, the creation of the Cell.

Additionally, I have added comments to clarify that query_value is deprecated. The configuration and synthesize should share the cells in the same way they share the columns.

3.

If we do what you propose, that would return to the previous status quo, where each circuit has a different cell manager.

4.

Because of the example in the wikipedia you might have inferred that the “Context class” (Car) should implement more features than the strategy interface (IBrake). That is never stated as the intention of the Strategy Pattern. It is just a choice of example.

5.

In this more extended article https://refactoring.guru/design-patterns/strategy

We can see that in its example, the context class (Navigator) only goal from the point of the client is to build a route, and then it wraps different ways of doing it (walking, by car…) in different strategies that implement the same interface.

It is not necessary	or condition or related at all that the context class have other actions beyond the one done in the strategies. It is an interface between the client and the strategies.

6.

The goal of the strategy pattern is NOT to wrap different behaviours in a single class with internal implementer for each behaviour, like Break, Accelerate, etc,,

The goal of the strategy pattern is to give the clients of a problem solver a single point of entry to different strategies to solve that problem.

So in our cases, the point is that all circuits will communicate with only one struct CellManager, and then the strategy used will be transparent to the circuit.

leolara commented:
@hero78119 I am sorry if I sounded rude 🙏
